# ETL example and exercise notebook

In this notebook the example data from the book and other data is loaded and examples of ETL is presented. Futhermore, there are exercises you can try out yourself.

## 1 Loading data into R

In this section we will load different types of data into R.

First we will load the data from the Books [Github page](https://github.com/jgendron/com.packtpub.intro.r.bi). You can see the raw data [here](https://raw.githubusercontent.com/jgendron/com.packtpub.intro.r.bi/master/Chapter1-ExtractTransformLoad/data/Ch1_bike_sharing_data.csv)

In [1]:
url <- "https://raw.githubusercontent.com/jgendron/com.packtpub.intro.r.bi/master/Chapter1-ExtractTransformLoad/data/Ch1_bike_sharing_data.csv"
bikeData <- read.csv(url)

The above command first store the url of the data in the a variable called `url`. Then, this variable is passed to the function `read.csv` that reads the data into the data frame `bikeData`. (So you can read csv files directly into R from the web without downloading the file first.)

You can see the first 6 lines of the data set by using the function `head` on `bikeData`:

In [2]:
head(bikeData)

datetime      season holiday workingday weather temp atemp  humidity
1 1/1/2011 0:00 1      0       0          1       9.84 14.395 81      
2 1/1/2011 1:00 1      0       0          1       9.02 13.635 80      
3 1/1/2011 2:00 1      0       0          1       9.02 13.635 80      
4 1/1/2011 3:00 1      0       0          1       9.84 14.395 75      
5 1/1/2011 4:00 1      0       0          1       9.84 14.395 75      
6 1/1/2011 5:00 1      0       0          2       9.84 12.880 75      
  windspeed casual registered count
1 0.0000    3      13         16   
2 0.0000    8      32         40   
3 0.0000    5      27         32   
4 0.0000    3      10         13   
5 0.0000    0       1          1   
6 6.0032    0       1          1

If you want to se another number of the first line, you can pass that number as the second argument to the `head` function. For instance, if you want to see the first 15 lines of the `bikeData` data frame, you can simply do:

In [3]:
head(bikeData, 15)

datetime       season holiday workingday weather temp  atemp  humidity
1  1/1/2011 0:00  1      0       0          1        9.84 14.395 81      
2  1/1/2011 1:00  1      0       0          1        9.02 13.635 80      
3  1/1/2011 2:00  1      0       0          1        9.02 13.635 80      
4  1/1/2011 3:00  1      0       0          1        9.84 14.395 75      
5  1/1/2011 4:00  1      0       0          1        9.84 14.395 75      
6  1/1/2011 5:00  1      0       0          2        9.84 12.880 75      
7  1/1/2011 6:00  1      0       0          1        9.02 13.635 80      
8  1/1/2011 7:00  1      0       0          1        8.20 12.880 86      
9  1/1/2011 8:00  1      0       0          1        9.84 14.395 75      
10 1/1/2011 9:00  1      0       0          1       13.12 17.425 76      
11 1/1/2011 10:00 1      0       0          1       15.58 19.695 76      
12 1/1/2011 11:00 1      0       0          1       14.76 16.665 81      
13 1/1/2011 12:00 1      0       0          1       17.22 21.210 77      
14 1/1/2011 13:00 1      0       0          2       18.86 22.725 72      
15 1/1/2011 14:00 1      0       0          2       18.86 22.725 72      
   windspeed casual registered count
1   0.0000    3     13          16  
2   0.0000    8     32          40  
3   0.0000    5     27          32  
4   0.0000    3     10          13  
5   0.0000    0      1           1  
6   6.0032    0      1           1  
7   0.0000    2      0           2  
8   0.0000    1      2           3  
9   0.0000    1      7           8  
10  0.0000    8      6          14  
11 16.9979   12     24          36  
12 19.0012   26     30          56  
13 19.0012   29     55          84  
14 19.9995   47     47          94  
15 19.0012   35     71         106

### *Exercise*

What do you think the `tail` function does? Try to read the help page by typing `?tail` and try using the `tail` function on the `bikeData` data frame below.

While the `head` function gives you an example of the data by showing you the first lines, the `str` function will provide you with more information about the data frame. Try it out:

In [ ]:
str(bikeData)

As you can see the `str` function tells you that `bikeData` is a `'data.frame'` and that it has 17379 rows and 12 columns. Moreover, for each column it tells you the name, the data type, and give you examples of the first couple of values. For instance, the `temp` variable is of numeric data type and the first couple of values are 9.84, 9.02, 9.02, ... Similar, `count` is a integer data type with the first couple of values 16, 40, 32, 13, ...

The datatime variable is a Factor variable as you can also see from above. It also tell you the number of levels of the factor variable. In this case the number of factors is the same as the number of observations/rows. This should make you suspicious. What you really want is the `datetime` variable to be of some datetime data type. To get the datetime column imported as basic text string instead of factor, we can tell R to read all string variables as strings instead of factors (- by default R read all string variables in as factors). To do this pass the argument `stringsAsFactors=FALSE` to the function reading in the data:

In [ ]:
bikeData <- read.csv(url, stringsAsFactors=FALSE)
str(bikeData)

As you can see the `datatime` column is now of type `chr` which means character (or text string).

### Loading in excel files

Let us try to import another data set. This time we will load an excel sheet from a local folder, that is from a folder in the current Azure Notebook library.

If you want to load in a file from you local computer you first have to upload the file to the Azure notebook library. So you should start by uploading the file "Webanalytics_data_example.xlsx" (Find this on the Moodle page for todays lecture) to your clone of the "BIBA 2018" library. To do this, from the Azure notebook library click the "+ New" and them "From computer". Click "Choose file" and then "Upload".

Before we can read in the file, we need a function that can read excel files. Such a function we can find in one of the packages the comes with the "tidyverse" collection of packages. This collection of packages is already installed on the Azure notebooks servers (if you work locally, you can install the packages using the command `install.packages("tidyverse")`). The particular package we need is called "readxl" and we load it with the following command:

In [4]:
library(readxl)

The package "readr" contains the function `read_excel` that we can try and read the file we just uploaded with:

In [8]:
webAnalyticsData <- read_excel("Webanalytics_data_example.xlsx")

Let us look at what we loaded in:

In [9]:
head(webAnalyticsData)

Here is some data sets as they might look when you download them from Google Analytics

Something seems weird here... If you open the file in Excel (try to do that), you will see that there are multiple sheets, and the one we loaded in is the first one. We want the second one, that is the sheet named "Dataset1". However, reading the help page for the function `read_excel` tell us that it is easily done by adding the argument `sheet = 2` or `sheet = Dataset1`. So try this:

In [10]:
webAnalyticsData <- read_excel("Webanalytics_data_example.xlsx", sheet = "Dataset1")
head(webAnalyticsData)

MediaChannel   Sessions PercentNewSessions NewUsers BounceRate
1 Organic_Search 76852    0.7840915          60259    0.3948368 
2 Social         59398    0.9433146          56031    0.5423752 
3 Paid_Search    31882    0.6663948          21246    0.6219183 
4 Direct         26348    0.7586534          19989    0.4594656 
5 Referral       18762    0.5305405           9954    0.2699606 
6 Display         6411    0.8502574           5451    0.8341912 
  PagesPerSession AvgSessionDuration Transactions Revenue  
1  7.225290       161.32634           989          94966.75
2  3.232348       111.79883            38           1692.29
3  4.461326        97.98413           425          21515.42
4  6.277820       162.81221           367          52054.79
5 12.226149       259.00032          1231         202558.88
6  2.359850        34.35470            34           7956.42

### *Exercise*

Use the `str` function to figure out what the different data types are of the different columns

### *Exercise*

Try to load the third sheet ("Dataset2") into a data frame called "webAnalyticsData2". (You might run into some problems. If so check the excel file and consider whether you need to skip some lines. Figure out how to do this by looking at the help page for the function `read_excel`. (If you manage to do it correctly, the cell below should give you a nice plot.)

In [11]:
# If you did the exercise correctly, this code should give you a nice plot
library(dplyr); library(ggplot2); webAnalyticsData2 %>% select(Dates = contains("Day"), Sessions) %>% ggplot(aes(Dates, Sessions)) + geom_line()


Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union



ERROR: Error in eval(lhs, parent, parent): object 'webAnalyticsData2' not found


## 2 Transforming data in R 

Let us try and do some transformation on the data we have just loaded into R.

### *Exercise*

Filter the `webAnalyticsData` data frame on the rows for which the bounce rate i bigger than 0.6.

(Bounce rate is defined as the percentage of users who left a website immediately after going there, in other words they most likely ended there by mistake or did not find the website interesting. Thus this will give us an idea of which MediaChannels provide the biggest fraction of useless visitors to the website.)

In [12]:
filter(webAnalyticsData, BounceRate > 0.6)

MediaChannel Sessions PercentNewSessions NewUsers BounceRate PagesPerSession
1 Paid_Search  31882    0.6663948          21246    0.6219183  4.461326       
2 Display       6411    0.8502574           5451    0.8341912  2.359850       
  AvgSessionDuration Transactions Revenue 
1 97.98413           425          21515.42
2 34.35470            34           7956.42

### *Exercise*

Filter the `webAnalyticsData` data frame on the rows fow which Transactions i bigger than 100 and Sessions is less than 30000

(This give us media channels with a decent amount of transactions given not to many sessions. Thus the are somewhat effective these media channels.)

In [13]:
filter(webAnalyticsData, Transactions > 100 & Sessions < 30000)

MediaChannel Sessions PercentNewSessions NewUsers BounceRate PagesPerSession
1 Direct       26348    0.7586534          19989    0.4594656   6.27782       
2 Referral     18762    0.5305405           9954    0.2699606  12.22615       
  AvgSessionDuration Transactions Revenue  
1 162.8122            367          52054.79
2 259.0003           1231         202558.88

### *Exercise*

We can create a more precise measure for the effectiveness of a media channel by calculating the precentage of sessions that ended in a transaction. That is for how many of the users sessions that the website had, did the user end up actually buying something. This is often also refered to as the conversion rate.

Use the `mutate` function to create a new column called `ConversionRate`, which is `Transactions/Sessions`. Which media had the highest conversion rate?

In [14]:
mutate(webAnalyticsData, ConversionRate = Transactions / Sessions)

MediaChannel   Sessions PercentNewSessions NewUsers BounceRate
1 Organic_Search 76852    0.7840915          60259    0.3948368 
2 Social         59398    0.9433146          56031    0.5423752 
3 Paid_Search    31882    0.6663948          21246    0.6219183 
4 Direct         26348    0.7586534          19989    0.4594656 
5 Referral       18762    0.5305405           9954    0.2699606 
6 Display         6411    0.8502574           5451    0.8341912 
7 Other             29    0.3103448              9    0.5862069 
8 Email             19    0.2105263              4    0.3157895 
  PagesPerSession AvgSessionDuration Transactions Revenue   ConversionRate
1  7.225290       161.32634           989          94966.75 0.0128688909  
2  3.232348       111.79883            38           1692.29 0.0006397522  
3  4.461326        97.98413           425          21515.42 0.0133304059  
4  6.277820       162.81221           367          52054.79 0.0139289510  
5 12.226149       259.00032          1231         202558.88 0.0656113421  
6  2.359850        34.35470            34           7956.42 0.0053033848  
7  2.448276        57.34483             0              0.00 0.0000000000  
8  3.368421       132.89474             0              0.00 0.0000000000

### *Exercise*

We might instead be interested in the absolute revenue generated by the different media channels. To see this in a easier readable format, use the `select` function to only select the columns `MediaChannel` and `Revenue`.

In [17]:
select(webAnalyticsData, MediaChannel, Revenue)

MediaChannel   Revenue  
1 Organic_Search  94966.75
2 Social           1692.29
3 Paid_Search     21515.42
4 Direct          52054.79
5 Referral       202558.88
6 Display          7956.42
7 Other               0.00
8 Email               0.00

### *Exercise*

To make it even easier to see the media channel with the highest revenue, arrange the data frame such that the media channel with the highest revenue is at the top. (Hint: You need to use the `select` and `arrange` function at the same time. Thus, you might need to make a nested function call, or store the output of one of the calls in a temporary variable.)

In [22]:
arrange(select(webAnalyticsData, MediaChannel, Revenue), desc(Revenue))
select(arrange(webAnalyticsData, desc(Revenue)), MediaChannel, Revenue)

MediaChannel   Revenue  
1 Referral       202558.88
2 Organic_Search  94966.75
3 Direct          52054.79
4 Paid_Search     21515.42
5 Display          7956.42
6 Social           1692.29
7 Other               0.00
8 Email               0.00

MediaChannel   Revenue  
1 Referral       202558.88
2 Organic_Search  94966.75
3 Direct          52054.79
4 Paid_Search     21515.42
5 Display          7956.42
6 Social           1692.29
7 Other               0.00
8 Email               0.00

### *Exercise*

We are interested in knowing the total revenue and conversion rate for paid and non-paid media. Thus we will group the media channels into paid and non-paid. Thus, first create a new column called `type` which takes on the value `Paid` for the media channels `Paid_Search`, `Display` and `Email`. For the other channels the value should be `Non-paid` (it is of course debatable whether to put social in paid or non-paid here). To do this use `mutate` and the following expression to define the `Type` column: `ifelse(MediaChannel %in% c("Paid_Search", "Display", "Email"), "Paid", "Non-paid")`. (Try to look up the help for the `ifelse` function and see if you can understand this code.)

Store the output of this exercise in a data frame called `webAnalyticsDatawType`.

In [24]:
webAnalyticsDatawType = mutate(webAnalyticsData, Type = ifelse(MediaChannel %in% c("Paid_Search", "Display", "Email"), "Paid", "Non-paid"))

In [25]:
webAnalyticsDatawType

MediaChannel   Sessions PercentNewSessions NewUsers BounceRate
1 Organic_Search 76852    0.7840915          60259    0.3948368 
2 Social         59398    0.9433146          56031    0.5423752 
3 Paid_Search    31882    0.6663948          21246    0.6219183 
4 Direct         26348    0.7586534          19989    0.4594656 
5 Referral       18762    0.5305405           9954    0.2699606 
6 Display         6411    0.8502574           5451    0.8341912 
7 Other             29    0.3103448              9    0.5862069 
8 Email             19    0.2105263              4    0.3157895 
  PagesPerSession AvgSessionDuration Transactions Revenue   Type    
1  7.225290       161.32634           989          94966.75 Non-paid
2  3.232348       111.79883            38           1692.29 Non-paid
3  4.461326        97.98413           425          21515.42 Paid    
4  6.277820       162.81221           367          52054.79 Non-paid
5 12.226149       259.00032          1231         202558.88 Non-paid
6  2.359850        34.35470            34           7956.42 Paid    
7  2.448276        57.34483             0              0.00 Non-paid
8  3.368421       132.89474             0              0.00 Paid

### *Exercise*

Use the data frame from the last exercise together with the `groupby` and `summerise` functions to create a data frame that shows the total revenue and total number of sessions for each to types `Paid` and `Non-paid`. (HInt: It is probably easiest to make a nested function call the with `group_by` function as the inner function.)

In [29]:
summarise(group_by(webAnalyticsDatawType, Type), totalRevenue = sum(Revenue), totalSessions = sum(Sessions))

Type     totalRevenue totalSessions
1 Non-paid 351272.71    181389       
2 Paid      29471.84     38312

### *Exercise*

Calculate the conversion rate for Paid and Non-paid media. (Hint: Use the `group_by` and `summarise` functions.)

In [33]:
df <- mutate(summarise(group_by(webAnalyticsDatawType, Type), totalTransactions = sum(Transactions), totalSessions = sum(Sessions)), ConversionRate = totalTransactions / totalSessions)
df

Type     totalTransactions totalSessions ConversionRate
1 Non-paid 2625              181389        0.01447166    
2 Paid      459               38312        0.01198058

### *Exercise*

Write the data frame of the last exercise to a csv file. Can you locate the file in Azure Notebooks and does it looks like you expected?

In [34]:
write.csv(df, "conversion_report.csv", row.names = FALSE)